File: Java/src/CoinChange.java

public class CoinChange {
	public int coinChange(int[] coins, int amount) {
		if(amount == 0) return 0;
		
		int [] dp = new int [amount+1];
		for(int i : coins) {
			if(i <= amount) dp[i] = 1;
		}
		
		for(int i = 1; i <= amount; i++) {
			if(dp[i] != 0) continue;
			int min = Integer.MAX_VALUE;
			for(int j : coins) {
				if(i - j > 0 && dp[i-j] != 0) {
					if(dp[i - j] < min) {
						min = dp[i-j];
					}
				}
			}
			if(min != Integer.MAX_VALUE) dp[i] = min + 1;
		}
		
		if(dp[amount] == 0) return -1;
		else return dp[amount];
        
    }
	
	public static void main(String[] args) {
		CoinChange cc = new CoinChange();
		int[] coins = {1, 2, 5};
		System.out.println(cc.coinChange(coins, 11));
	}

}

File: Python/CoinChange.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 0
        max = 100000000
        for i in range(1, amount+1):
            dp[i] = max
            for c in coins:
                if i - c >= 0:
                    dp[i] = min(dp[i], dp[i-c]+1)
        if dp[amount] >= max:
            return -1
        return dp[amount]
        


